server/worker/worker.go

```diff
@@ -47,37 +48,59 @@ func (w *Worker) Register(jobs ...Job) {
 	}
 }
 
-func (w *Worker) QueueJob(ctx context.Context, job *fleet.Job) (*fleet.Job, error) {
-	job.State = fleet.JobStateQueued
-	return w.ds.NewJob(ctx, job)
+// QueueJob inserts a job to be processed by the worker for the job processor
+// identified by the name (e.g. "jira"). The args value is marshaled as JSON
+// and provided to the job processor when the job is executed.
+func QueueJob(ctx context.Context, ds fleet.Datastore, name string, args interface{}) (*fleet.Job, error) {
+	argsJSON, err := json.Marshal(args)
+	if err != nil {
+		return nil, ctxerr.Wrap(ctx, err, "marshal args")
+	}
+	job := &fleet.Job{
+		Name:  name,
+		Args:  (*json.RawMessage)(&argsJSON),
+		State: fleet.JobStateQueued,
+	}
+
+	return ds.NewJob(ctx, job)
 }
 
 // ProcessJobs processes all queued jobs.
 func (w *Worker) ProcessJobs(ctx context.Context) error {
+	const maxNumJobs = 100
 
-	// process jobs until there are none left
-	maxNumJobs := 100
+	// process jobs until there are none left or the context is cancelled
+	var startID uint
 	for {
 		jobs, err := w.ds.GetQueuedJobs(ctx, maxNumJobs)
 		if err != nil {
-			return fmt.Errorf("get jobs: %w", err)
+			return ctxerr.Wrap(ctx, err, "get queued jobs")
 		}
 
 		if len(jobs) == 0 {
 			break
 		}
 
 		for _, job := range jobs {
+			select {
+			case <-ctx.Done():
+				return ctxerr.Wrap(ctx, ctx.Err(), "context done")
+			default:
+			}
+
 			log := kitlog.With(w.log, "job_id", job.ID)
 
+			if job.ID == startID {
+				level.Debug(log).Log("msg", "back to initial job of that batch, stopping until next cron execution")
+				return nil
+			}
```

Comment:
I don't think this works as intended. It would only work if the job that fails was the first job in that batch. You should use a map of seen jobs instead
```
seen := make(map[uint]struct{}) // store seen jobs
```
this might also require additional changes to `GetQueuedJobs` so that it accepts a starting ID. Originally, it was meant to process all jobs until the state is failure. This would no longer be the case...